### PR TITLE
DFU: progresive image erase

### LINKED ISSUE #8734: USB: DFU timeout on nRF52840 due to long flash erase
Attempt of downloading new firmware with dfu-util on nRF52840 results in timeout due to flash erase operation taking too much time (flash_img_buffered_write() called from dfu_flash_write() in usb_dfu.c.

This problem could be solved by fragmenting whole flash erase operation into smaller page erases when needed.

Working USB driver for testing is available on this fork/branch:  https://github.com/pawelzadrozniak/zephyr/tree/nrfx_usbd3 .

### PR DESCRIPTION
**Pull request by nvlsianpu**
Patch adds option for progressive erase of firmware image.
Using this flash is erased as necessary when receiving
new firmware, instead of erasing the whole image slot at once.
This is necessary on some hardware that has long erase times,
to prevent long wait times at the beginning of the DFU process.

usb/class/usb_dfu: support progressive image erase

Disable bulk slot image erase when progressive erase is on.
Erase of image bank is performed by image collection procedure
progressively.

For some SoC like nRF52840 bulk erase take too much time, which
cause timeout on tool on existing system out-of-the-box (dfu-util
or just by downloading package from sourceforge for Windows).

~~resolves #8734~~

### CODE REVIEW

File: subsys/usb/class/usb_dfu.c [hunk 1]
```diff
@@ -731,11 +731,15 @@ static void dfu_work_handler(struct k_work *item)
 
 	switch (dfu_data_worker.worker_state) {
 	case dfuIDLE:
+/* If progressive erase is enabled, then erase take place while */
+/* image collection, so not erase whole bank at DFU beginning   */
```
Comment (Qbicz) [inline subsys/usb/class/usb_dfu.c]: it could be a single multi-line comment

**Comment by nvlsianpu** [comment]
@Qbicz I spotted that the last erase (while flushing the buffer) has  the bug you showed me. Will fix this.

**Comment by nvlsianpu** [comment]
@Qbicz it is fixed.

**Comment by Qbicz** [comment]
Hi @nvlsianpu, I tested the progressive image erase and it works. I was able to perform USB DFU on nRF52840 even with the dfu-util with 5 second timeout.

File: subsys/dfu/img_util/flash_img.c [hunk 2]
```diff
@@ -52,6 +57,63 @@ static bool flash_verify(const struct flash_area *fa, off_t offset,
...
+	return rc;
+}
+
+/**
+ * Erase the image slot progressively
+ *
+ * This function erases a flash page to which offset belongs if only this page
+ * wasn't erased before.
+ *
+ * @param[in] ctx contex of the image collection process.
```
Comment (Qbicz) [inline subsys/dfu/img_util/flash_img.c]: s/contex/context

File: subsys/dfu/img_util/flash_img.c [hunk 3]
```diff
@@ -52,6 +57,63 @@ static bool flash_verify(const struct flash_area *fa, off_t offset,
...
+					      sector.fs_size);
+			if (rc) {
+				LOG_ERR("Error %d while erasing sector", rc);
+			}
+		}
+	}
+
+	return rc;
+}
+
```
Comment (Qbicz) [inline subsys/dfu/img_util/flash_img.c]: excessive empty line

**Comment by Qbicz** [comment]
For me this PR can be merged, only the minor changes need to be done + rebase on master.

File: subsys/dfu/img_util/flash_img.c (hunk 3 above)
Comment (nvlsianpu) [reply subsys/dfu/img_util/flash_img.c]: prefer to keep, for me it improve legibility

**Comment by nvlsianpu** [comment]
@Qbicz revisit

File: subsys/usb/class/usb_dfu.c [hunk 4]
```diff
@@ -731,11 +731,17 @@ static void dfu_work_handler(struct k_work *item)
 
 	switch (dfu_data_worker.worker_state) {
 	case dfuIDLE:
+/*
+ * If progressive erase is enabled, then erase take place while
```
Comment (Qbicz) [inline subsys/usb/class/usb_dfu.c]:
```suggestion
 * If progressive erase is enabled, erase takes place during
```

File: subsys/usb/class/usb_dfu.c [hunk 5]
```diff
@@ -731,11 +731,17 @@ static void dfu_work_handler(struct k_work *item)
...
+ * image collection, so not erase whole bank at DFU beginning
```
Comment (Qbicz) [inline subsys/usb/class/usb_dfu.c]:
```suggestion
 * image collection, in order not to erase whole bank at DFU beginning
```

**Review by Qbicz — APPROVED** [review]

**Comment by Qbicz** [comment]
@finikorg @jfischer-phytec-iot can you review?

**Comment by tormodvolden** [comment]
If this is meant to solve issues with dfu-util, it seems like the wrong fix or at the best a complicated workaround. When performing long operations like a bulk erase, the device should signal to the host (dfu-util) the time before it should resume polling the device, through the bwPollTimeout field. Long operations are only to be initiated after the host has requested the status through the DFU_GETSTATUS request (following the DFU_DNLOAD request that typically encodes an erase operation), whose reply includes the poll timeout value. Please see the DFU 1.1 specification at https://www.usb.org/sites/default/files/DFU_1.1.pdf

There is a 5 second timeout value in dfu-util for the DFU request _USB transaction_, but any long operation should not hold up this transaction.

BTW, there is also a 5 second poll timeout default value in dfu-util, but it is only applied as a last resort for known, broken devices that doesn't fill in the bwPollTimeout field correctly.

**Comment by tormodvolden** [comment]
Looking at your code, here is what you will need to fix:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/class/usb_dfu.c#L396

You have mistakenly hardcoded the bwPollTimeout to 256 ms. This value should be set dynamically, according to the time the device will need for the previously requested operation, e.g. how long it will be unresponsive after a bulk erase. It will typically be set differently for read/write operations and RAM/flash.

**Comment by jfischer-phytec-iot** [comment]
@tormodvolden tanks for the hint

>You have mistakenly hardcoded the bwPollTimeout to 256 ms. This value should be set dynamically, according to the time the device will need for the previously requested operation, e.g. how long it will be unresponsive after a bulk erase. It will typically be set differently for read/write operations and RAM/flash.

I will change it.

**Comment by nvlsianpu** [comment]
@tormodvolden Thanks for investigation and explanation.
I will keep this PR main content (progresive erase) as it was requested for other adoption as well.
Question is: should I remove it dfu class commit?

**Comment by tormodvolden** [comment]
At least the commit messages should be changed IMO, there is no problem with long erase times in the beginning of a DFU process, now that the device implements the bwPollTimeout correctly. So that part of the motivation for this pull request goes out. And no other reason is mentioned :)

**Comment by Qbicz** [comment]
While the real fix for https://github.com/zephyrproject-rtos/zephyr/issues/8734 is https://github.com/zephyrproject-rtos/zephyr/pull/15330, I would still like to merge this PR, of course with changed commit message.
Progressive image erase has some benefits: 
- it prevents user from waiting long time before update process is started.
- according to @nvlsianpu in bulk erase whole image slot is erased. For progressive erase only the pages that need to be erased

Moreover, this is an optional feature. I would be happy to have it available to improve user experience.

**Review by Qbicz — COMMENTED** [review]
Can we change first commit message to

```
Patch adds option for progressive erase of firmware image.
When using this, flash is erased as necessary when receiving
new firmware, instead of erasing the whole image slot at once.
This is useful on some hardware (like nRF52840) that has
long erase times, to prevent long wait times at the beginning
of the DFU process.
```

And the second commit message to
```
Disable bulk slot image erase when progressive erase is on.
Erase of image bank is performed by image collection procedure
progressively.
```

i.e. remove information about this fixing an issue with dfu-util.

**Comment by nvlsianpu** [comment]
@Qbicz Thanks a lot for your comment. Of course I will rehearsed messages.

**Comment by tormodvolden** [comment]
IMO, whether dfu-util or not, "useful [...] to prevent long wait times" isn't a much of a technical motivation though. Or is it just for the user experience that the DFU process seems to be hold up with no visible progress during some seconds?

OTOH, what I see as a huge advantage here (if I understand correctly) is that you can download to parts of the flash without erasing other parts.

**Comment by nvlsianpu** [comment]
@tormodvolden 
Progressive erase feature is required by some applications we have to build (is not necessary USB DFU, just better User experience) class. It is optional and disabled by default.

**Comment by nvlsianpu** [comment]
@tormodvolden @jfischer-phytec-iot commit message updated.

**Review by jfischer-no — APPROVED** [review]

**Comment by jfischer-phytec-iot** [comment]
@carlescufi @nvlsianpu @Qbicz I will revert ea177e7, as it was not thought to end and does not work properly. So this should be merged asap and backported to 1.14.

**Pull request merged**